### ISSUE TITLE
Look into OneDNN integration

### ISSUE THREAD

**Issue by rengolin** [plaidml/tpp-mlir#209]
Intel's [OneDNN](https://github.com/oneapi-src/oneDNN) library has support for many CPUs and GPUs and could allow us to connect to an existing end-to-end pipeline just like [IREE](https://github.com/plaidml/tpp-mlir/issues/167) or [OpenXLA](https://github.com/plaidml/tpp-mlir/issues/208).

However, there are a few issues with it:
1. It only [validates against RedHat Linux](https://github.com/oneapi-src/oneDNN#validated-configurations), restricting to GCC/Clang version 9, which are too old. We can build those configurations on the cluster, but harder on our local machines, not to mention most people who do ML work use Ubuntu and GCC/Clang 12 or later.
2. It supports Tensorflow and PyTorch, but [require additional extensions to work](https://github.com/oneapi-src/oneDNN#applications-enabled-with-onednn), which may complicate our end-to-end prototype, if we want to keep the delta low.

Still, would be good to look at it and make sure we can adapt to it as well.

@nhasabni

**Comment by rengolin** [comment]
We may not get to do these, so I'll close it in order to avoid issue pollution. If we want to do this in the future, we can reopen.

**Comment by rengolin** [comment]
Reopening this one as we should be in a reasonable place to translate high-level linalg named ops (whole tensors) with existing OneDNN functions. For this to work, we'll need to have OneDNN as a (library) dependency and get those in our `tpp-run` to include in the runtime.

Since the deprecation of libxsmm-dnn, we could replace those benchmarks completely (and the build/lib dependency) with OneDNN kernels for "perfect shapes".

@alheinecke @nhasabni @hfp

**Comment by rengolin** [comment]
To get a PoC here, we should separate the work into a few tasks:
- [ ] Import OneDNN as an optional dependency to the project, at least for the kernel libraries
  - [ ] Check if OneDNN already has an MLIR mechanism (dialect?) to lower to its calls
  - [ ] If not, we have to decide to encode that as functions/runtime hooks or a new dialect
- [ ] Lower some MLP models into MLIR (ex. via PyTorch in https://github.com/plaidml/mlir-generator)
- [ ] Find the patterns that match existing OneDNN kernels (GEMMs, fused ops) and try to match them directly before tile & fuse
  - [ ] May need to kludge tensor code through veneer functions, bufferization and cleanups to match OneDNN signatures
  - [ ] Or create a shim runtime layer like we have for XSMM to marshal the arguments and return values
- [ ] Once we can reach peak performance, we look at a more complex example, like MHA / Transformer

We can do this for CPU first and when `tpp-run` is ready to run more complex patterns (may need https://github.com/llvm/llvm-project/pull/65539), we do the same for GPU. 

@adam-smnk / @hfp: would be nice to have a time-frame, even if uncertain at the time.

@Jianhui-Li any insight into this task?

@BenBrock would be nice to have your input here, too.

@chelini once we have MHA lowering, we can compare your compiler work with the OneDNN version and see how far we're from actual peak performance.

**Comment by chelini** [comment]
I can take on how to integrate.

**Comment by adam-smnk** [comment]
I think we could ramp up basic OneDNN runner integration in Q4.
That would have to cover extra dependencies, build integration, runtime hooks (probably most work if needed), basic native examples, and baseline benchmarking.

**Comment by chelini** [comment]
> I can take on how to integrate.

#716 started the integration.

**Comment by Jianhui-Li** [comment]
@retonym has done some early work integrating oneDNN Graph API to IREE.  You need to use the Graph AI to access advanced fusion like MHA.  We can certainly follow up with this.  

@npatel @Zantares

**Comment by adam-smnk** [comment]
> some early work integrating oneDNN Graph API to IREE.

@retonym @Jianhui-Li could you point us to this work?
So far, I haven't found any OneDNN -> MLIR logic.

In case we need to bring up sth ourselves, I'd propose a small oneapi/onednn dialect as the non-BLAS OneDNN API gets more complex. Also, some pros that come to my head:
* just like Xsmm dialect - same dialect role and approach
* makes lowering more progressive - nicer design, (should be) easier to maintain
* representative example - a more proper showcase on how to integrate with MLIR
* potentially more scalable - it should be easier to shape MHA using OneDNN IR compared to trying to glue everything in runtime wrapper

**Comment by rengolin** [comment]
Another thing I discussed with @Jianhui-Li is how we match MHAs and all its variants. If we lower from StableHLO/Torch/TCP into Linalg, we'd do some pattern match to raise it again to OneDNN calls, which is not good.

If this dialect comes from above, say when OneDNN graph lowers StableHLO to Linalg, what we receive is a mix of OneDNN invocations (function calls, some dialect) and Linalg. We can only look at Linalg, so we won't break the OneDNN stuff.

However, this creates a packing problem. If OneDNN has its own packing, it will pack and unpack inside the routine, and we'll pack and unpack around the routines. Without a way to communicate the packing decisions, we'll add a lot of unnecessary data movements.

Our first PoC shouldn't have to worry about these issues, for example, we can assume we only lower to OneDNN or we only lower to XSMM, but our second goal is to make sure we can do both, and that's where multi-level cost models and heuristics need to enter the picture.

---

PS: Specifically for the packing problem, if OneDNN lowers their functions with pack/unpack before/after the call, we can still see the packs and propagate the same packs throughout the code. This would be a easier middle ground between no information and full graph analysis.

**Comment by rengolin** [comment]
@chelini, now we need a way to plug in more generic calls to gemms. OneDNN supports large kernels, so we're not concerned with tile sized GEMMs here. Means it makes no sense to build a small set of variants with different dims.

I suggest we do something very generic here, which even works for dynamic shapes:
1. Take the size of each dim, store in an SSA value
2. Take the inputs and cast to the most generic form (strided memref)
3. Call DNN functions with those as arguments

The serialisation of those arguments seem trivially similar to me (before understanding the whole complexity), so I'd propose to common that up in a helper function or use a single dispatch function (with switches for the kernels). The former is cleaner, but the latter is quicker to prototype. Since this isn't going to be a final design, it should be fine either way.

**Comment by alheinecke** [comment]
We should not use any pack/unpack for oneDNN and not any outer loops etc. as you say. I also agree with @adam-smnk  that an IR with named ops of the oneDNN primitives would be very beneficial as we cannot use "sgemm" (as discussed in #716 ). The group concept we are thinking of for LinAlg Tiles, could be then also applied to full LinAlg ops / oneDNN we can call in the RT oneDNN Graph. E.g. then we can compare MHA best possible: Library vs. Code-Gen.

**Comment by rengolin** [comment]
Agreed, let's not worry about packing just yet. This is not a problem now, because we're only looking into whole kernel functions to compare them to our compiler's output. 

It will be a problem once we start looking into larger graphs, especially new models that don't have well defined OneDNN kernels implemented. But we can cross that bridge when we get there.

---

Long Read: To be clear, here's what I meant:

If some operations are lowered into OneDNN and others not, the ones still in linalg could be packed by us, while the ones in OneDNN could be packed by them (or not packed at all). Intermixing large kernel calls with linalg operations will forbid pack propagation.

I don't know if OneDNN packs its inputs, but the boundary condition is still a valid concern when interoperating libraries at different levels, because not all decisions are taken by the compiler, and it may not be possible to communicate them across libraries.

A simple example:
```mlir
func.func main (%arg0, ...) {
  %0 = call onednn_fused_brgemm(%arg0, ...) : tensor<original_shape>
  %1 = linalg.matmul (%0, ...)
  %2 = call onednn_fused_brgemm(%1, ...) : tensor<original_shape>
  %3 = linalg.matmul (%2, ...)
  return %3
}
```

Tile and fuse may be tempted to pack `%1` and `%3`, but it won't be able to propagate it through the OneDNN call in `%2`.

A more complex example:
```mlir
func.func main (...) {
  // OneDNN fused kernel
  %0 = linag.batch_reduced_matmul ...
  %1 = linalg.add (%0, ...
  %2 = linalg.max (%1, %cst_0)

  // Unrecognised kernel
  %3 = linalg.generic (%2) { mul, add } // possible to lower to brgemm
  %4 = linalg.generic (%3) { add } // tail EW
  return %4
}
```

Say we lower the first to OneDNN and the second remain in linalg, then we run pack&tile&fuse pass:
```mlir
// Example of its C impl
func.func onednn_fused_brgemm(%arg0, ...) { : tensor<original_shape>
  %0 = pack { tileX, tileY } %arg0
  tile loop {
    %1 = brgemm
    %2 = add
    %3 = relu
  }
  %4 = unpack
  return %4
}

func.func main (...) {
  // OneDNN fused kernel
  %0 = call onednn_fused_brgemm(...) : tensor<original_shape>

  // Tilable kernel: tileW/tileZ not necessarily the same as tileX/tileY
  %1 = tensor.pack { tileW, tileZ } %0
  %2 = parallel_for {
    %tile = tensor.extract_slice
    // This can be an XSMM fused brgemm
    %i_0 = linalg.generic (%tile) { mul, add }
    %i_1 = linalg.generic (%i_0) { add }
    tensor.parallel_insert_slice (%i_0)
  }
  %3 = tensor.unpack (%2)
  return %3
}
```

Not only we could have many packs and unpacks, but they could be different decisions (written by different groups with different strategies). Ideally, we'd communicate the packing decisions across libraries and compilers, and be able to select the right (pre-packed if necessary) kernels so that the compiler can propagate the shapes.

```mlir
func.func main (...) {
  // OneDNN fused kernel, rewriter exposes tile choices, returns packed shape
  %0 = call onednn_fused_brgemm(...) { tileX, tileY } : tensor<packed_shape>

  // Tilable kernel, no need for pack
  %2 = parallel_for {
    %tile = tensor.extract_slice
    // This can be an XSMM fused brgemm
    %i_0 = linalg.generic (%tile) { mul, add }
    %i_1 = linalg.generic (%i_0) { add }
    tensor.parallel_insert_slice (%i_0)
  }
  // Still needs to unpack
  %3 = tensor.unpack (%2)
  return %3
}
```

**Comment by rengolin** [comment]
Preliminary work done in #716 and #723. We won't be working on this in the near future, so I'll close this issue.
